File: java/main/ch/softappeal/yass/core/remote/session/Reconnector.java
package ch.softappeal.yass.core.remote.session;

import ch.softappeal.yass.util.Check;

import java.util.concurrent.Executor;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.TimeUnit;

/**
 * Provides proxies surviving reconnects.
 */
public class Reconnector<S extends Session> extends ProxyDelegate<S> {

    @FunctionalInterface public interface Connector {
        /**
         * @throws Exception note: will be ignored
         */
        void connect(SessionFactory sessionFactory) throws Exception;
    }

    /**
     * @param executor used once; must interrupt it's threads to terminate reconnects (use {@link ExecutorService#shutdownNow()})
     */
    public final void start(
        final Executor executor, final long delaySeconds, final long intervalSeconds,
        final SessionFactory sessionFactory, final Connector connector
    ) {
        Check.notNull(sessionFactory);
        Check.notNull(connector);
        @SuppressWarnings("unchecked") final SessionFactory reconnectorSessionFactory = connection -> {
            final Session session = sessionFactory.create(connection);
            session((S)session);
            return session;
        };
        executor.execute(() -> {
            try {
                TimeUnit.SECONDS.sleep(delaySeconds);
            } catch (final InterruptedException ignore) {
                return;
            }
            while (!Thread.interrupted()) {
                if (!connected()) {
                    session(null);
                    try {
                        connector.connect(reconnectorSessionFactory);
                    } catch (final Exception ignore) {
                        // empty
                    }
                }
                try {
                    TimeUnit.SECONDS.sleep(intervalSeconds);
                } catch (final InterruptedException ignore) {
                    return;
                }
            }
        });
    }

    public final void start(
        final Executor executor, final long intervalSeconds,
        final SessionFactory sessionFactory, final Connector connector
    ) {
        start(executor, 0L, intervalSeconds, sessionFactory, connector);
    }

}

File: kotlin/yass/main/ch/softappeal/yass/remote/session/Reconnector.kt
package ch.softappeal.yass.remote.session

import ch.softappeal.yass.*
import java.lang.reflect.*
import java.util.concurrent.*

private fun isConnected(session: Session?): Boolean =
    (session != null) && !session.isClosed

abstract class ProxyDelegate<S : Session> {
    @Volatile
    private var _session: S? = null

    val session: S
        get() {
            if (!isConnected) throw SessionClosedException()
            return _session!!
        }

    protected fun setSession(session: S?) {
        _session = session
    }

    val isConnected get() = isConnected(_session)

    @OnlyNeededForJava
    fun <C : Any> proxy(contract: Class<C>, proxyGetter: (session: S) -> C): C =
        proxy(
            contract,
            InvocationHandler { _, method, arguments -> invoke(method, proxyGetter(session), args(arguments)) }
        )

    inline fun <reified C : Any> proxy(noinline proxyGetter: (session: S) -> C): C =
        proxy(C::class.java, proxyGetter)
}

/** Provides proxies surviving reconnects. */
open class Reconnector<S : Session> : ProxyDelegate<S>() {
    /**
     * [executor] is called once; must interrupt it's threads to terminate reconnects.
     * Thrown exceptions of [connector] will be ignored.
     */
    @JvmOverloads
    fun start(
        executor: Executor,
        intervalSeconds: Long, sessionFactory: SessionFactory, delaySeconds: Long = 0,
        connector: (sessionFactory: SessionFactory) -> Unit
    ) {
        require(intervalSeconds >= 1)
        require(delaySeconds >= 0)
        val reconnectorSessionFactory = {
            val session = sessionFactory()
            @Suppress("UNCHECKED_CAST") setSession(session as S)
            session
        }
        executor.execute {
            try {
                TimeUnit.SECONDS.sleep(delaySeconds)
            } catch (e: InterruptedException) {
                return@execute
            }
            while (!Thread.interrupted()) {
                if (!isConnected) {
                    setSession(null)
                    try {
                        connector(reconnectorSessionFactory)
                    } catch (ignore: Exception) {
                    }
                }
                try {
                    TimeUnit.SECONDS.sleep(intervalSeconds)
                } catch (e: InterruptedException) {
                    return@execute
                }
            }
        }
    }
}
